src/html/classic/ng/src/web/pages/extras/trashcanpage.js

```diff
@@ -0,0 +1,493 @@
+import TableBody from '../../components/table/body.js';
+import TableData from '../../components/table/data.js';
+import TableRow from '../../components/table/row.js';
+import TableHead from '../../components/table/head.js';
+import TableHeader from '../../components/table/header.js';
+import {Col} from 'glamorous';
+
+import CollectionList from 'gmp/collection/collectionlist.js';
+
+import Layout from '../../components/layout/layout.js';
+import Section from '../../components/section/section.js';
+
+import InnerLink from '../../components/link/innerlink.js';
+import LinkTarget from '../../components/link/target.js';
+
+import Button from '../../components/form/button.js';
+import HelpIcon from '../../components/icon/helpicon.js';
+
+import Loading from '../../components/loading/loading.js';
+
+import {map, is_defined} from 'gmp/utils.js';
+import PropTypes from '../../utils/proptypes.js';
+import withContext from '../../utils/withContext.js';
+
+
+const ToolBarIcons = () => (
+  <HelpIcon
+    page="trashcan"
+    title={_('Help: Trashcan')}
+  />
+);
+
+const EmptyTrashButton = ({onClick}) => (
+  <Layout align="right">
+    <Button
+      onClick={onClick}
+    >{_('Empty Trash')}</Button>
+  </Layout>
+);
+
+class Trashcan extends React.Component {
+
+  constructor(...args) {
+    super(...args);
+    this.state = {
+      trash_data: undefined,
+      contents_table: [],
+    };
+
+    this.createContentRow = this.createContentRow.bind(this);
+    this.createContentsTable = this.createContentsTable.bind(this);
+    this.getTrash = this.getTrash.bind(this);
+    this.handleEmptyTrashcan = this.handleEmptyTrashcan.bind(this);
+    this.handleDelete = this.handleDelete.bind(this);
+    this.handleRestore = this.handleRestore.bind(this);
+  }
+
+  componentDidMount() {
+      this.getTrash();
+  }
+
+  getTrash() {
+    const {gmp} = this.props;
+    const data = Promise.all([
+      gmp.trashcan.readTrashcan().then(response => {
+        const trash_data = response.data;
+        const contents_table = this.createContentsTable(trash_data);
+        this.setState({trash_data, contents_table});
+      }),
+    ]);
+    return data;
+  }
+
+  handleRestore(entity) {
+    const {gmp} = this.props;
+    gmp.trashcan.restore(entity).then(this.getTrash);
+  }
+
+  handleDelete(entity) {
+    const {gmp} = this.props;
+    gmp.trashcan.delete(entity).then(this.getTrash);
+  }
+
+  handleEmptyTrashcan() {
+    const {gmp} = this.props;
+    gmp.trashcan.emptyTrashcan().then(this.getTrash);
+  }
+
+  createContentRow(type, count) {
+    return (
+      <TableRow key={type}>
+        <TableData>
+          <InnerLink to={type}>{type}</InnerLink>
+        </TableData>
+        <TableData>
+          {count}
+        </TableData>
+      </TableRow>
+    );
+  }
+
+  createContentsTable(trash) {
+    return (
+      <TableBody>
+        {this.createContentRow('Agents',
+          trash.get_agents_response.agent_count.toString())}
```

Comment:
You should improve your Command class that the response already contains a better data structure. You need the counts therefore it should only contain the counts here: tash.get_xyz_response is unnecessary. tash.agent_count would be the better api. Maybe even name it tash.agents_count.

Reply:
Btw. toString() isn't needed at all. It's already a string. Also converting from number to string wouldn't be necessary here. And converting to string would be better with '' + value

Reply:
Done